---	
title: "Intersect VarScan and Strelka"	
author: "Stacy Hung"	
date: "November 25, 2019"	
output: html_document	
---	

Integration of VarScan, Strelka, and MuTect predictions for 14 trios of tumor vs. pooled blood and matched normals vs. pooled blood.  Adds tumor vs. matched normal calls made previously.

## Load libraries
 	
```{r}
library(dplyr)        # filter	
library(plyr)         # revalue
library(tidyr)        # separate	
library(DataCombine)  # find and replace
library(reshape2)     # colsplit
```

## Define samples to be excluded and global variables

```{r}

# read in data that contains information on tumor content and tissue type of the normal
sample.data <- read.table("/Volumes/shung/projects/gzl_targeted/data/gzl_targeted-T_N-samples.txt", sep = "\t", header = TRUE, fill = TRUE, as.is = TRUE)

# only keep samples that were sequenced as part of validation cohort
sample.data <- filter(sample.data, sample.data$result == "validation")

# extract only required columns
sample.data <- sample.data[, c("tumor.tumor_id", "tumor.tissue_type.v2", "tumor_content", "tumor.mean_target_coverage")]
colnames(sample.data) <- c("tumor.id", "tumor.tissue_type", "tumor_content", "tumor.mean_target_coverage")

# exclude LMD samples that have < 40X coverage or < 100X for all other samples
sample.data.1 <- filter(sample.data, tumor.tissue_type == "FFPE_LMD" & as.numeric(tumor.mean_target_coverage) > 40)
sample.data.2 <- filter(sample.data, tumor.tissue_type != "FFPE_LMD" & as.numeric(tumor.mean_target_coverage) > 100)
sample.data <- rbind(sample.data.1, sample.data.2)

# effects to exclude
effects.exclude <- c("intron_variant", "intragenic_variant", "", "sequence_feature")

# panel genes
genes.panel <- read.table("/Volumes/shung/projects/gzl_exomes/data/target_panel_Agilent_genes.txt", header=FALSE)
genes.panel <- as.vector(genes.panel$V1)

# filter for protein coding biotypes
biotype.keeps <- c("IG_C_gene", "IG_D_gene", "IG_gene", "IG_J_gene", "IGL_V_gene", "IG_M_gene", "IG_V_gene",
                   "IG_Z_gene", "nonsense_mediated_decay", "nontranslating_CDS", "non_stop_decay",
                   "polymorphic_pseudogene", "TR_C_gene", "TR_D_gene", "TR_J_gene", "protein_coding")

# define thresholds for all samples EXCEPT for tumors paired to LMD normals
MAX_SOMATIC_PVAL = 0.05
MIN_VAR_READS_TUMOR = 3

# germline filters where tumor content is <20% (temporary) --> remove variants with VAF between 45-55% or >90%
GERMLINE_AF_HETERO_LOW = 45   # lower bound for heterozygous germline variant
GERMLINE_AF_HETERO_HIGH = 55  # upper bound for heterozygous germline variant
GERMLINE_AF_HOMO_LOW = 90      # lower bound for homozygous germline variant

# miniumum depth at position:
MIN_DEPTH = 20;

# minimum number of variant reads in normal
MAX_VAR_READS_NORMAL = 2;

# define normal groups
normals.pooled <- c("pooledN1", "pooledN2")

# define "tumor" samples that should be analyzed
samples <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/data/sample_pairs.txt", sep = "\t", header = FALSE)
tumors <- unique(samples$V1)
normals <- unique(samples$V2)
```

## VARSCAN

```{r}
varScan.snvs.pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/varscan/all_snvs.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
varScan.indels.pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/varscan/all_indels.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
varScan.snvs.matchedN <- read.table("/Volumes/shung/projects/gzl_targeted/all_samples/varscan/all_snvs.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
varScan.indels.matchedN <- read.table("/Volumes/shung/projects/gzl_targeted/all_samples/varscan/all_indels.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)

# specify type of mutation
varScan.indels.pooledN$type <- "indel"
varScan.snvs.pooledN$type <- "snv"
varScan.indels.matchedN$type <- "indel"
varScan.snvs.matchedN$type <- "snv"

# combine snvs and indels
varScan.calls <- rbind(varScan.indels.pooledN, varScan.snvs.pooledN, varScan.indels.matchedN, varScan.snvs.matchedN)
rm(varScan.indels.pooledN)
rm(varScan.snvs.pooledN)
rm(varScan.indels.matchedN)
rm(varScan.snvs.matchedN)

# rename columns	
colnames(varScan.calls) <- c("sample", "chr", "pos", "external_id", "ref", "alt", "filter", 
                             "gt_normal", "gq_normal", "normal.depth", "normal.ref_reads", "normal.var_reads",
                             "normal.allele_freq", "normal.depth4", "tumor.gt", "tumor.gq", "tumor.depth",
                             "tumor.ref_reads", "tumor.var_reads", "tumor.allele_freq", "tumor.depth4", 
                             "depth", "somatic", "somatic_status", "gpv", "somatic_p_value", "allele", 
                             "effect", "impact", "gene", "feature", "feature_id", "biotype", "exon_rank", 
                             "hgvs_cdna", "hgvs_protein", "cds_pos", "cds_len", "AA_pos", "AA_len", 
                             "LOF_gene", "LOF_geneid", "LOF_num_Tr", "LOF_perc", "type")

# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
varScan.calls$tumor.id <- gsub("(.*)\\_(.*)","\\1", varScan.calls$sample)
# similarly create a normal_id column
varScan.calls$normal.id <- gsub("(.*)\\_(.*)","\\2", varScan.calls$sample)

# filter for analysis tumors
varScan.calls <- filter(varScan.calls, varScan.calls$tumor.id %in% tumors)

# remove "%" from the allele frequency	
varScan.calls$normal.allele_freq <- gsub("(.*)\\%(.*)","\\1", varScan.calls$normal.allele_freq)	
varScan.calls$tumor.allele_freq <- gsub("(.*)\\%(.*)","\\1", varScan.calls$tumor.allele_freq)	


#################################
# filters applicable to VarScan #
#################################

# apply obvious filters: somatic == true ~ somatic_status == 2
varScan.calls <- filter(varScan.calls, varScan.calls$somatic == "true")

# filter for somatic p-value
varScan.calls <- filter(varScan.calls, as.numeric(somatic_p_value) < MAX_SOMATIC_PVAL)

# strand bias filter
varScan.calls <-	
  separate(data = varScan.calls,	
           col = tumor.depth4,	
           into = c("tumor.ref_reads.forward", "tumor.ref_reads.reverse",
                  "tumor.var_reads.forward", "tumor.var_reads.reverse"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)	

# NB: there is a huge number of variants exhibiting strand bias, so number of mutations is severely reduced after this step
varScan.calls <- filter(varScan.calls, 
                        !(as.numeric(tumor.var_reads.forward) > as.numeric(0.9*as.numeric(tumor.var_reads))) & 
                          !(as.numeric(tumor.var_reads.reverse) > as.numeric(0.9*as.numeric(tumor.var_reads))))

# update actual tumor depth
varScan.calls$tumor.depth <- varScan.calls$tumor.ref_reads.forward + varScan.calls$tumor.ref_reads.reverse +
  varScan.calls$tumor.var_reads.forward + varScan.calls$tumor.var_reads.reverse

# remove columns for strand bias check since they are no longer needed
varScan.calls$tumor.ref_reads.forward <- NULL
varScan.calls$tumor.ref_reads.reverse <- NULL
varScan.calls$tumor.var_reads.forward <- NULL
varScan.calls$tumor.var_reads.reverse <- NULL
varScan.calls$tumor.depth4 <- NULL


###################################
# filters applicable to all tools #
###################################

# 1. min required reads in tumor
varScan.calls <- filter(varScan.calls, as.numeric(tumor.var_reads) >= MIN_VAR_READS_TUMOR)

# 2. normal VAF < 5% of tumor VAF
# 3. At most 2 variant reads in the normal
varScan.calls <- filter(varScan.calls,
                             (as.numeric(normal.allele_freq) < 0.05*as.numeric(tumor.allele_freq)) &
                             (as.numeric(normal.var_reads) <= MAX_VAR_READS_NORMAL)
                           )

# 4. filter for protein-altering effects
varScan.calls <- filter(varScan.calls, !(varScan.calls$effect %in% effects.exclude)) 

# 5. filter out low impact variants
varScan.calls <- filter(varScan.calls, varScan.calls$impact != "LOW")

# 6. only include variants annotated on Ensembl transcript
varScan.calls <- unique(varScan.calls[grep("^ENST", varScan.calls$feature_id), ])

# 7. filter for panel genes
varScan.calls <- filter(varScan.calls, varScan.calls$gene %in% genes.panel) 

# 8. filter out non protein coding biotypes
varScan.calls <- filter(varScan.calls, varScan.calls$biotype %in% biotype.keeps)


# remove unneeded columns and rearrange as necessary	
keeps <- c("tumor.id", "normal.id", "type", "chr", "pos", "gene", "ref", "alt", 
           "normal.depth", "normal.var_reads", "normal.allele_freq", 
           "tumor.depth", "tumor.var_reads", "tumor.allele_freq", 
           "effect", "impact", "biotype", "feature_id", "external_id")	
varScan.calls <- varScan.calls[keeps]	

# merge T-N data with SNVs
varScan.calls <- merge(varScan.calls, sample.data, by = "tumor.id", all.x = TRUE)


# remove duplicates (e.g. due to multiple effects)
varScan.calls <- unique(varScan.calls)	

varScan.calls$combine = as.character(paste(	
  varScan.calls$tumor.id, 	
  varScan.calls$type,
  varScan.calls$chr, 	
  varScan.calls$pos, 	
  varScan.calls$ref,	
  varScan.calls$alt,	
  sep = "."	
  ))

```

## MUTECT

```{r}
mutect.snvs.pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/mutect/all_snvs.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
mutect.snvs.matchedN <- read.table("/Volumes/shung/projects/gzl_targeted/all_samples/mutect/all_snvs.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)

mutect.snvs.matchedN$type = "snv"
mutect.snvs.pooledN$type = "snv" 

mutect.snvs <- rbind(mutect.snvs.matchedN, mutect.snvs.pooledN)
rm(mutect.snvs.matchedN)
rm(mutect.snvs.pooledN)

# rename columns
colnames(mutect.snvs) <- c("sample", "chr", "pos", "external_id", "ref", 
          "alt", "filter", "normal.gt", "normal.var_reads", "normal.base_qual", 
          "normal.depth", "normal.allele_freq", "normal.somatic_status", "tumor.gt", "tumor.var_reads", 
          "tumor.base_qual", "tumor.depth", "tumor.allele_freq", "somatic_status", "somatic", "VT", "allele", 
          "effect", "impact", "gene", "feature", "feature_id", "biotype", "exon_rank", "hgvs_cdna", "hgvs_protein", 
          "cds_pos", "cds_len", "AA_pos", "AA_len", "LOF_gene", "LOF_geneid", "LOF_num_Tr", "LOF_perc", "type")
                         
# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
mutect.snvs$tumor.id <- gsub("(.*)\\_(.*)","\\1", mutect.snvs$sample)
# similarly create a normal_id column
mutect.snvs$normal.id <- gsub("(.*)\\_(.*)","\\2", mutect.snvs$sample)

# filter for analysis tumors
mutect.snvs <- filter(mutect.snvs, mutect.snvs$tumor.id %in% tumors)

# convert allele frequencies to percent instead of fraction (to be consistent with VarScan)
mutect.snvs$normal.allele_freq <- mutect.snvs$normal.allele_freq*100
mutect.snvs$tumor.allele_freq <- mutect.snvs$tumor.allele_freq*100

# merge T-N data with SNVs to apply additional filters
mutect.snvs <- merge(mutect.snvs, sample.data, by = "tumor.id", all.x = TRUE)

################################
# filters applicable to Mutect #
################################

# filter out low-quality bases (BQ < 20)
mutect.snvs <- filter(mutect.snvs, mutect.snvs$tumor.base_qual >= 20)

# unlike in VarScan, depth at variant position is not filtered; filter for min 20X coverage
mutect.snvs <- filter(mutect.snvs, as.numeric(tumor.depth) >= MIN_DEPTH)

# extract tumor variant reads
mutect.snvs <-	
  separate(data = mutect.snvs,	
           col = tumor.var_reads,	
           into = c("tumor.ref_reads", "tumor.var_reads"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)	
mutect.snvs <-	
  separate(data = mutect.snvs,	
           col = normal.var_reads,	
           into = c("normal.ref_reads", "normal.var_reads"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)	

# remove unneeded columns and rearrange as necessary	
keeps <- c("tumor.id", "normal.id", "type", "chr", "pos", "gene", "ref", "alt", 
           "normal.depth", "normal.var_reads", "normal.allele_freq", 
           "tumor.depth", "tumor.var_reads", "tumor.allele_freq", 
           "effect", "impact", "biotype", "feature_id", "external_id")	
mutect.snvs <- mutect.snvs[, keeps]	

###################################
# filters applicable to all tools #
###################################

# 1. min required reads in tumor
mutect.snvs <- filter(mutect.snvs, as.numeric(tumor.var_reads) >= MIN_VAR_READS_TUMOR)

# 2. normal VAF < 5% of tumor VAF
# 3. At most 2 variant reads in the normal
mutect.snvs <- filter(mutect.snvs,
                             (as.numeric(normal.allele_freq) < 0.05*as.numeric(tumor.allele_freq)) &
                             (as.numeric(normal.var_reads) <= MAX_VAR_READS_NORMAL)
                           )

# 4. filter for protein-altering effects
mutect.snvs <- filter(mutect.snvs, !(mutect.snvs$effect %in% effects.exclude)) 

# 5. filter out low impact variants
mutect.snvs <- filter(mutect.snvs, mutect.snvs$impact != "LOW")

# 6. only include variants annotated on Ensembl transcript
mutect.snvs <- unique(mutect.snvs[grep("^ENST", mutect.snvs$feature_id), ])

# 7. filter for panel genes
mutect.snvs <- filter(mutect.snvs, mutect.snvs$gene %in% genes.panel) 

# 8. filter out non protein coding biotypes
mutect.snvs <- filter(mutect.snvs, mutect.snvs$biotype %in% biotype.keeps)

mutect.snvs$combine = as.character(paste(
  mutect.snvs$tumor.id,
  mutect.snvs$type,
  mutect.snvs$chr,
  mutect.snvs$pos,
  mutect.snvs$ref,
  mutect.snvs$alt,
  sep = "."
))

```

## STRELKA

Extracting alelle frequencies for strelka SNVs:

refCounts = value in "GEN.TUMOR..XU" column; where REF = X {A, C, G, T}
altCounts = value in "GEN.TUMOR..YU" column; where ALT = Y {A, C, G, T}
tier1RefCounts = first comma-delimited value from refCounts
tier1AltCounts = first comma-delimited value from altCounts
somatic allele frequency = tier1AltCounts / (tier1AltCounts + tier1RefCounts)
  
Extracting allele frequencies for strelka indels:

tier1RefCounts = first comma-delimited value in TAR
tier1AltCounts = first comma-delimited value in TIR
somatic allele frequency = tier1AltCounts / (tier1AltCounts + tier1RefCounts)
  

```{r}
strelka.snvs.pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/strelka/all_snvs.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
strelka.indels.pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/strelka/all_indels.snpEff_canonical.dbsnp_annotated.cosmic_annotated.txt", sep = "\t", header = TRUE, fill = TRUE)
strelka.snvs.matchedN <- read.table("/Volumes/shung/projects/gzl_targeted/all_samples/strelka/all.snvs.txt", sep = "\t", header = TRUE, fill = TRUE)
strelka.indels.matchedN <- read.table("/Volumes/shung/projects/gzl_targeted/all_samples/strelka/all.indels.txt", sep = "\t", header = TRUE, fill = TRUE)

# label type of mutation
strelka.snvs.pooledN$type <- "snv"
strelka.snvs.matchedN$type <- "snv"
strelka.indels.pooledN$type <- "indel"
strelka.indels.matchedN$type <- "indel"

# combine datasets, but keep SNVs and indels as separate datasets since they need to be parsed differently
strelka.snvs <- rbind(strelka.snvs.matchedN, strelka.snvs.pooledN)
strelka.indels <- rbind(strelka.indels.matchedN, strelka.indels.pooledN)
rm(strelka.snvs.matchedN)
rm(strelka.snvs.pooledN)
rm(strelka.indels.matchedN)
rm(strelka.indels.pooledN)

# create a tumor_id column based on the sample id (format is <tumor_id>_<normal_id>) - e.g. GE0556B_GE0556-N
strelka.snvs$tumor.id <- gsub("(.*)\\_(.*)","\\1", strelka.snvs$SAMPLE)
strelka.indels$tumor.id <- gsub("(.*)\\_(.*)","\\1", strelka.indels$SAMPLE)
# similarly create a normal_id column
strelka.snvs$normal.id <- gsub("(.*)\\_(.*)","\\2", strelka.snvs$SAMPLE)
strelka.indels$normal.id <- gsub("(.*)\\_(.*)","\\2", strelka.indels$SAMPLE)

# filter for analysis tumors
strelka.snvs <- filter(strelka.snvs, strelka.snvs$tumor.id %in% tumors)
strelka.indels <- filter(strelka.indels, strelka.indels$tumor.id %in% tumors)

# extract columns for filtering: variant tumor reads, allele frequency, effect, gene

# columns of unknown meaning:
#   "GEN.NORMAL..FDP"
#   "GEN.TUMOR..FDP" 

## Extract somatic variant allele frequencies from strelka snvs - 5 steps:
## Note that the steps are the same for calculating tumor and normal (except column names are different)

# 1) refCounts = value in "GEN.TUMOR..XU" column; where REF = X {A, C, G, T}
strelka.snvs <- strelka.snvs %>%
  mutate(refCounts = ifelse(as.character(REF) == "A", as.character(strelka.snvs$GEN.TUMOR..AU),
                            ifelse(as.character(REF) == "C", as.character(strelka.snvs$GEN.TUMOR..CU),
                                   ifelse(as.character(REF) == "G", as.character(strelka.snvs$GEN.TUMOR..GU),
                                          as.character(strelka.snvs$GEN.TUMOR..TU)))))
strelka.snvs <- strelka.snvs %>%
  mutate(normal.refCounts = ifelse(as.character(REF) == "A", as.character(strelka.snvs$GEN.NORMAL..AU),
                            ifelse(as.character(REF) == "C", as.character(strelka.snvs$GEN.NORMAL..CU),
                                   ifelse(as.character(REF) == "G", as.character(strelka.snvs$GEN.NORMAL..GU),
                                          as.character(strelka.snvs$GEN.NORMAL..TU)))))

# 2) altCounts = value in "GEN.TUMOR..YU" column; where ALT = Y {A, C, G, T}
strelka.snvs <- strelka.snvs %>%
  mutate(altCounts = ifelse(as.character(ALT) == "A", as.character(strelka.snvs$GEN.TUMOR..AU),
                            ifelse(as.character(ALT) == "C", as.character(strelka.snvs$GEN.TUMOR..CU),
                                   ifelse(as.character(ALT) == "G", as.character(strelka.snvs$GEN.TUMOR..GU),
                                          as.character(strelka.snvs$GEN.TUMOR..TU)))))
strelka.snvs <- strelka.snvs %>%
  mutate(normal.altCounts = ifelse(as.character(ALT) == "A", as.character(strelka.snvs$GEN.NORMAL..AU),
                            ifelse(as.character(ALT) == "C", as.character(strelka.snvs$GEN.NORMAL..CU),
                                   ifelse(as.character(ALT) == "G", as.character(strelka.snvs$GEN.NORMAL..GU),
                                          as.character(strelka.snvs$GEN.NORMAL..TU)))))

# 3) tier1RefCounts = first comma-delimited value from refCounts
strelka.snvs <-	
  separate(data = strelka.snvs,	
           col = refCounts,	
           into = c("tier1RefCounts", "tier2RefCounts"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)
strelka.snvs$tier2RefCounts <- NULL

strelka.snvs <-	
  separate(data = strelka.snvs,	
           col = normal.refCounts,	
           into = c("normal.tier1RefCounts", "normal.tier2RefCounts"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)
strelka.snvs$normal.tier2RefCounts <- NULL

# 4) tier1AltCounts = first comma-delimited value from altCounts
strelka.snvs <-	
  separate(data = strelka.snvs,	
           col = altCounts,	
           into = c("tier1AltCounts", "tier2AltCounts"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)
strelka.snvs$tier2AltCounts <- NULL

strelka.snvs <-	
  separate(data = strelka.snvs,	
           col = normal.altCounts,	
           into = c("normal.tier1AltCounts", "normal.tier2AltCounts"),	
           sep = ",",	convert = TRUE,
           remove = FALSE)
strelka.snvs$normal.tier2AltCounts <- NULL

# 5) somatic allele frequency = tier1AltCounts / (tier1AltCounts + tier1RefCounts)
strelka.snvs$tumor.allele_freq <- 
  ( strelka.snvs$tier1AltCounts / (strelka.snvs$tier1AltCounts + strelka.snvs$tier1RefCounts) ) * 100
strelka.snvs$normal.allele_freq <- (strelka.snvs$normal.tier1AltCounts / (strelka.snvs$normal.tier1AltCounts + strelka.snvs$normal.tier1RefCounts) ) * 100

# create intuitive columns to represent (tumor/normal) depth and (tumor/normal) variant depth:
strelka.snvs$tumor.depth <- strelka.snvs$tier1AltCounts + strelka.snvs$tier1RefCounts
strelka.snvs$tumor.var_depth <- strelka.snvs$tier1AltCounts
strelka.snvs$normal.depth <- strelka.snvs$normal.tier1AltCounts + strelka.snvs$normal.tier1RefCounts
strelka.snvs$normal.var_depth <- strelka.snvs$normal.tier1AltCounts


## Extract somatic variant allele frequencies from strelka indels - 3 steps:

# 1) tier1RefCounts = first comma-delimited value in TAR
strelka.indels <-
  separate(data = strelka.indels,
           col = GEN.TUMOR..TAR,
           into = c("tier1RefCounts", "tier2RefCounts"),
           sep = ",", convert = TRUE,
           remove = FALSE)
strelka.indels$tier2RefCounts <- NULL

strelka.indels <-
  separate(data = strelka.indels,
           col = GEN.NORMAL..TAR,
           into = c("normal.tier1RefCounts", "normal.tier2RefCounts"),
           sep = ",", convert = TRUE,
           remove = FALSE)
strelka.indels$normal.tier2RefCounts <- NULL

# 2) tier1AltCounts = first comma-delimited value in TIR
strelka.indels <-
  separate(data = strelka.indels,
           col = GEN.TUMOR..TIR,
           into = c("tier1AltCounts", "tier2AltCounts"),
           sep = ",", convert = TRUE,
           remove = FALSE)
strelka.indels$tier2AltCounts <- NULL

strelka.indels <-
  separate(data = strelka.indels,
           col = GEN.NORMAL..TIR,
           into = c("normal.tier1AltCounts", "normal.tier2AltCounts"),
           sep = ",", convert = TRUE,
           remove = FALSE)
strelka.indels$normal.tier2AltCounts <- NULL

# 3) somatic allele frequency = tier1AltCounts / (tier1AltCounts + tier1RefCounts)
strelka.indels$tumor.allele_freq <- 
  (strelka.indels$tier1AltCounts / (strelka.indels$tier1AltCounts + strelka.indels$tier1RefCounts) )*100

strelka.indels$normal.allele_freq <- 
  (strelka.indels$normal.tier1AltCounts / (strelka.indels$normal.tier1AltCounts + strelka.indels$normal.tier1RefCounts) )*100

# create intuitive columns to represent (tumor/normal) depth and (tumor/normal) variant depth:
strelka.indels$tumor.depth <- strelka.indels$tier1AltCounts + strelka.indels$tier1RefCounts
strelka.indels$tumor.var_depth <- strelka.indels$tier1AltCounts
strelka.indels$normal.depth <- strelka.indels$normal.tier1AltCounts + strelka.indels$normal.tier1RefCounts
strelka.indels$normal.var_depth <- strelka.indels$normal.tier1AltCounts

# extract only columns of interest so that we can combine snvs and indels
keeps <- c("tumor.id", "normal.id", "type", "CHROM", "POS", "ANN....GENE", "REF", "ALT",
                 "ANN....HGVS_C", "ANN....HGVS_P", "normal.depth", "normal.var_depth", "normal.allele_freq",
                 "tumor.depth", "tumor.var_depth", "tumor.allele_freq", "ANN....EFFECT", "ANN....IMPACT",
                 "ANN....BIOTYPE", "ID", "ANN....FEATUREID")
strelka.indels <- strelka.indels[, keeps]
strelka.snvs <- strelka.snvs[, keeps]

# combine snvs and indels into a single calls dataframe
strelka.calls <- rbind(strelka.snvs, strelka.indels)
rm(strelka.indels)	
rm(strelka.snvs)

# rename columns so they are more intuitive:
colnames(strelka.calls) <- c("tumor.id", "normal.id", "type", "chr", "pos", "gene", "ref", "alt",
                             "hgvs_cdna", "hgvs_protein", "normal.depth", "normal.var_reads", "normal.allele_freq",
                             "tumor.depth", "tumor.var_reads", "tumor.allele_freq", "effect", "impact", "biotype",
                             "external_id", "feature_id")

#################################
# filters applicable to Strelka #
#################################

# unlike in VarScan, depth at variant position is not filtered; filter for min 20X coverage
strelka.calls <- filter(strelka.calls, as.numeric(tumor.depth) >= MIN_DEPTH)

###################################
# filters applicable to all tools #
###################################

# 1. min required reads in tumor
strelka.calls <- filter(strelka.calls, as.numeric(tumor.var_reads) >= MIN_VAR_READS_TUMOR)

# 2. normal VAF < 5% of tumor VAF
# 3. At most 2 variant reads in the normal
strelka.calls <- filter(strelka.calls,
                             (as.numeric(normal.allele_freq) < 0.05*as.numeric(tumor.allele_freq)) &
                             (as.numeric(normal.var_reads) <= MAX_VAR_READS_NORMAL)
                           )

# 4. filter for protein-altering effects
strelka.calls <- filter(strelka.calls, !(strelka.calls$effect %in% effects.exclude)) 

# 5. filter out low impact variants
strelka.calls <- filter(strelka.calls, strelka.calls$impact != "LOW")

# 6. only include variants annotated on Ensembl transcript
strelka.calls <- unique(strelka.calls[grep("^ENST", strelka.calls$feature_id), ])

# 7. filter for panel genes
strelka.calls <- filter(strelka.calls, strelka.calls$gene %in% genes.panel) 

# 8. filter out non protein coding biotypes
strelka.calls <- filter(strelka.calls, strelka.calls$biotype %in% biotype.keeps)

strelka.calls$combine = as.character(paste(
  strelka.calls$tumor.id,
  strelka.calls$type,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))

# merge T-N data with SNVs to apply additional filters
strelka.calls <- merge(strelka.calls, sample.data, by = "tumor.id", all.x = TRUE)

strelka.calls$combine = as.character(paste(
  strelka.calls$tumor.id,
  strelka.calls$type,
  strelka.calls$chr,
  strelka.calls$pos,
  strelka.calls$ref,
  strelka.calls$alt,
  sep = "."
))
```

## Get all variants

Include variants that are identified in:
  (i) VarScan + MuTect (snvs only)
  (ii) VarScan + Strelka (snvs, indels)
  (iii) MuTect + Strelka (snvs only)
  (iv) all three
  (v) unique to individual tools

```{r}
# retrieve specific overlaps
calls.varscan_and_mutect_and_strelka <- intersect(intersect(unique(varScan.calls$combine), unique(mutect.snvs$combine)),
                                                  unique(strelka.calls$combine)
                                                  )
calls.varscan_and_mutect_only <- intersect(unique(varScan.calls$combine), unique(mutect.snvs$combine))
calls.varscan_and_strelka_only <- intersect(unique(varScan.calls$combine), unique(strelka.calls$combine))
calls.mutect_and_strelka_only <- intersect(unique(mutect.snvs$ combine), unique(strelka.calls$combine))
#calls.varscan_only
#calls.mutect_only

# now label the "union" table
union.keeps <- c("tumor.id", "normal.id", "type", "chr", "pos", "gene", "ref", "alt", "combine")
varscan.union <- varScan.calls[, union.keeps]
mutect.union <- mutect.snvs[, union.keeps]
strelka.union <- strelka.calls[, union.keeps]
calls.union <- unique(rbind(varscan.union, mutect.union, strelka.union))

# label by which datasets combine is present in

# check if key is present in the individual datasets
calls.union <- calls.union %>% mutate(datasets = ifelse(combine %in% varScan.calls$combine, "varscan", 
                           ifelse(combine %in% strelka.calls$combine, "strelka", "mutect")))
         
# add more granular labelling
calls.union <- calls.union %>% 
  mutate(datasets = ifelse(combine %in% calls.varscan_and_mutect_only, "varscan, mutect", datasets))
calls.union <- calls.union %>% 
  mutate(datasets = ifelse(combine %in% calls.varscan_and_strelka_only, "varscan, strelka", datasets))
calls.union <- calls.union %>% 
  mutate(datasets = ifelse(combine %in% calls.mutect_and_strelka_only, "mutect, strelka", datasets))
calls.union <- calls.union %>% 
  mutate(datasets = ifelse(combine %in% calls.varscan_and_mutect_and_strelka, "varscan, mutect, strelka", datasets))
calls.union <- unique(calls.union)

to.merge <- c("normal.depth", "normal.var_reads", "normal.allele_freq", 
              "tumor.depth", "tumor.var_reads", "tumor.allele_freq",
              "effect", "impact", "biotype", "external_id", "combine")

# add other fields depending on which tool
varscan.calls.to.merge <- varScan.calls[, to.merge]
strelka.calls.to.merge <- strelka.calls[, to.merge]
mutect.calls.to.merge <- mutect.snvs[, to.merge]

# now arbitrarily split up the calls into varscan-, mutect- and strelka- (in that order) annotatable:
# we do not want to include any variants that are predicted by only 1 tool
varscan.datasets <- c("varscan, strelka", 
                      "varscan, mutect")
mutect.datasets <- c("mutect, strelka")
strelka.datasets <- c("varscan, mutect, strelka")

calls.union.varscan.temp <- filter(calls.union, calls.union$datasets %in% varscan.datasets)
calls.union.mutect.temp <- filter(calls.union, calls.union$datasets %in% mutect.datasets)
calls.union.strelka.temp <- filter(calls.union, calls.union$datasets %in% strelka.datasets)

# now add annotation to individual datasets and combine again
calls.all.varscan <- merge(calls.union.varscan.temp, varscan.calls.to.merge, by = "combine")
calls.all.mutect <- merge(calls.union.mutect.temp, mutect.calls.to.merge, by = "combine")
calls.all.strelka <- merge(calls.union.strelka.temp, strelka.calls.to.merge, by = "combine")

calls.all <- rbind(calls.all.varscan, calls.all.mutect, calls.all.strelka)
#calls.all$combine <- NULL

# merge meta data with variant calls
#calls.all <- merge(calls.all, sample.data, by = "tumor.id", all.x = TRUE)

# sort the calls by case, chr, then position	
calls.all <- unique(arrange(calls.all, tumor.id, chr, pos))

# add identifier column
pairs.T_N <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/data/pairs.T_N.txt", header = FALSE)
pairs.T_pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/data/pairs.T_pooledN.txt", header = FALSE)
pairs.N_pooledN <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/data/pairs.N_pooledN.txt", header = FALSE)
pairs.T_N$analysis <- "tumor_v_normal"
pairs.T_pooledN$analysis <- "tumor_v_pooledN"
pairs.N_pooledN$analysis <- "normal_v_pooledN"
pairs <- rbind(pairs.T_N, pairs.T_pooledN, pairs.N_pooledN)
colnames(pairs) <- c("pair", "analysis")

calls.all$pair <- paste(calls.all$tumor.id, calls.all$normal.id, sep = "_")
calls.all <- merge(calls.all, pairs, by = "pair")

# write out the calls for downstream interpretation
#write.table(calls.all, "/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/calls.all.txt", sep = "\t", row.names = FALSE, quote = FALSE)
```

## Preparing for germline filtering

```{r}
library(dplyr)

# to avoid having to re-run everything, just load the last version of the integrated calls
#calls.all <- read.table("/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/calls.all.txt", sep = "\t", header = TRUE)

pooled_analyses <- c("tumor_v_pooledN", "normal_v_pooledN")

# extract calls where "tumor" has been paired against a pooled N and matched N
calls.v_matchedN <- filter(calls.all, !(calls.all$analysis %in% pooled_analyses))
calls.v_pooledN <- filter(calls.all, calls.all$analysis %in% pooled_analyses)

```

## Germline VAF filtering *VERSION 1* - filter out VAF 45-55% and >90%

```{r}
# redefine germline VAF thresholds
GERMLINE_AF_HETERO_LOW <- 45
GERMLINE_AF_HETERO_HIGH <- 55
GERMLINE_AF_HOMO_LOW <- 90

calls.v_pooledN.germline_VAF_filtered <- filter(calls.v_pooledN, 
                        ( (as.numeric(tumor.allele_freq) < GERMLINE_AF_HETERO_LOW) | 
                             (as.numeric(tumor.allele_freq) > GERMLINE_AF_HETERO_HIGH & 
                                as.numeric(tumor.allele_freq) < GERMLINE_AF_HOMO_LOW)
                           
))
calls.all.germline_VAF_filtered <- rbind(calls.v_matchedN, calls.v_pooledN.germline_VAF_filtered)

calls.all <- calls.all.germline_VAF_filtered
```

## Germline VAF filtering *VERSION 2* - filter out VAF 40-60% and >90%

```{r}
# redefine germline VAF thresholds
GERMLINE_AF_HETERO_LOW <- 40
GERMLINE_AF_HETERO_HIGH <- 60
GERMLINE_AF_HOMO_LOW <- 90

calls.v_pooledN.germline_VAF_filtered <- filter(calls.v_pooledN, 
                        ( (as.numeric(tumor.allele_freq) < GERMLINE_AF_HETERO_LOW) | 
                             (as.numeric(tumor.allele_freq) > GERMLINE_AF_HETERO_HIGH & 
                                as.numeric(tumor.allele_freq) < GERMLINE_AF_HOMO_LOW)
                           
))
calls.all.germline_VAF_filtered <- rbind(calls.v_matchedN, calls.v_pooledN.germline_VAF_filtered)

calls.all <- calls.all.germline_VAF_filtered
```

## Germline VAF filtering *VERSION 3* - filter out VAF 40-60% and >90% but KEEP if associated with ids from COSMIC only

```{r}
# redefine germline VAF thresholds
GERMLINE_AF_HETERO_LOW <- 40
GERMLINE_AF_HETERO_HIGH <- 60
GERMLINE_AF_HOMO_LOW <- 90

# extract calls that are < 40%, or between 60-90% VAF
calls.v_pooledN.germline_VAF_filtered <- filter(calls.v_pooledN, 
                        ( (as.numeric(tumor.allele_freq) < GERMLINE_AF_HETERO_LOW) | 
                             (as.numeric(tumor.allele_freq) > GERMLINE_AF_HETERO_HIGH & 
                                as.numeric(tumor.allele_freq) < GERMLINE_AF_HOMO_LOW)
                           
))

# now extract calls that between 40-60% or >90% VAF 
calls.v_pooled.germline_VAF <- filter(calls.v_pooledN, 
                                      ( as.numeric(tumor.allele_freq) >= GERMLINE_AF_HETERO_LOW &
                                          as.numeric(tumor.allele_freq) <= GERMLINE_AF_HETERO_HIGH ) | 
                                        ( as.numeric(tumor.allele_freq) >= GERMLINE_AF_HOMO_LOW )
                                      )
# extract calls from above that also have COSMIC ids
calls.v_pooled.germline_VAF.cosm_id <- filter(calls.v_pooled.germline_VAF, 
                                              grepl("COSM", calls.v_pooled.germline_VAF$external_id))
# extract calls from the above dataset that do not have an rs id
calls.v_pooled.germline_VAF.cosm_id.no_rsid <- filter(calls.v_pooled.germline_VAF.cosm_id,
                                                      !grepl("rs", calls.v_pooled.germline_VAF.cosm_id$external_id) )

# now define the final set of calls (germline VAF filtered + germline VAF with COSMIC only ids)
calls.all.germline_VAF_filtered.COSM_only_id <- rbind(calls.v_pooledN.germline_VAF_filtered,
                                                      calls.v_pooled.germline_VAF.cosm_id.no_rsid)
calls.all <- calls.all.germline_VAF_filtered.COSM_only_id
```

## Germline filtering based on VAF removing variants with a unique rs id

```{r}
# Step 1 - extract variants from pool datasets that *do not* have a DBSNP ID
calls.v_pooledN.no_rsid <- filter(calls.v_pooledN, !grepl("rs", calls.v_pooledN$external_id))

# Step 2 - extract variants from pool datasets that have a DBSNP ID
calls.v_pooledN.rsid <- filter(calls.v_pooledN, grepl("rs", calls.v_pooledN$external_id))

# Step 3 - extract variants from DBSNP subset that also have a COSMIC id
calls.v_pooledN.rs_and_COSM_id <- filter(calls.v_pooledN.rsid, grepl("COSM", calls.v_pooledN.rsid$external_id))
rm(calls.v_pooledN.rsid)

calls.v_pooledN.rsid_filtered <- rbind(calls.v_pooledN.no_rsid, calls.v_pooledN.rs_and_COSM_id)

calls.all.rsid_filtered <- rbind(calls.v_matchedN, calls.v_pooledN.rsid_filtered)

calls.all <- calls.all.rsid_filtered
```


## Second layer of combining - create union of calls, and column "analysis" indicating in which paired analyses, the variant appears in (e.g. "tumor vs. normal, tumor vs. pooledN")

```{r}
calls.T_v_N <- filter(calls.all, calls.all$analysis == "tumor_v_normal")          
calls.T_v_pooledN <- filter(calls.all, calls.all$analysis == "tumor_v_pooledN")   
calls.N_v_pooledN <- filter(calls.all, calls.all$analysis == "normal_v_pooledN")

#################################################################################
# for normal vs. pooledN variants, we need to change the "tumor id" so that the variants can be compared at the case level

calls.N_v_pooledN$tumor.id <- gsub("-C", "", calls.N_v_pooledN$tumor.id) # e.g. for normals like "BCC53-C"
calls.N_v_pooledN$pair <- gsub("-C", "", calls.N_v_pooledN$pair)

calls.N_v_pooledN$tumor.id <- gsub("C$", "", calls.N_v_pooledN$tumor.id) # e.g. for normals like "GZ184C"
calls.N_v_pooledN$pair <- gsub("C_", "_", calls.N_v_pooledN$pair)

# map BM1803125 to BCC25 (case id)
calls.N_v_pooledN$tumor.id <- gsub("BM1803125", "BCC25", calls.N_v_pooledN$tumor.id)
calls.N_v_pooledN$pair <- gsub("BM1803125", "BCC25", calls.N_v_pooledN$pair)

# update ids missing a dash (in normal - e.g. GZ180) (since all case ids have a dash - e.g. GZ-180)
calls.N_v_pooledN$tumor.id <- gsub("GZ086", "GZ-086", calls.N_v_pooledN$tumor.id)
calls.N_v_pooledN$tumor.id <- gsub("GZ092", "GZ-092", calls.N_v_pooledN$tumor.id)
calls.N_v_pooledN$tumor.id <- gsub("GZ149", "GZ-149", calls.N_v_pooledN$tumor.id)
calls.N_v_pooledN$tumor.id <- gsub("GZ180", "GZ-180", calls.N_v_pooledN$tumor.id)
calls.N_v_pooledN$tumor.id <- gsub("GZ184", "GZ-184", calls.N_v_pooledN$tumor.id)

calls.N_v_pooledN$pair <- gsub("GZ086", "GZ-086", calls.N_v_pooledN$pair)
calls.N_v_pooledN$pair <- gsub("GZ092", "GZ-092", calls.N_v_pooledN$pair)
calls.N_v_pooledN$pair <- gsub("GZ149", "GZ-149", calls.N_v_pooledN$pair)
calls.N_v_pooledN$pair <- gsub("GZ180", "GZ-180", calls.N_v_pooledN$pair)
calls.N_v_pooledN$pair <- gsub("GZ184", "GZ-184", calls.N_v_pooledN$pair)

# recreate "combine" key for normal vs. pooled normal variants
calls.N_v_pooledN$combine = as.character(paste(
  calls.N_v_pooledN$tumor.id,
  calls.N_v_pooledN$type,
  calls.N_v_pooledN$chr,
  calls.N_v_pooledN$pos,
  calls.N_v_pooledN$ref,
  calls.N_v_pooledN$alt,
  sep = "."
))

#################################################################################

# retrieve specific overlaps
combine.T_v_N.T_v_pooledN.N_v_pooledN <- intersect(intersect(unique(calls.T_v_N$combine), unique(calls.T_v_pooledN$combine)),
                                                 unique(calls.N_v_pooledN)) # 0 
combine.T_v_N.T_v_pooledN <- intersect(unique(calls.T_v_N$combine), unique(calls.T_v_pooledN$combine)) # 1133
combine.T_v_N.N_v_pooledN <- intersect(unique(calls.T_v_N$combine), unique(calls.N_v_pooledN$combine)) # 0
combine.T_v_pooledN.N_v_pooledN <- intersect(unique(calls.T_v_pooledN$combine), unique(calls.N_v_pooledN$combine)) # 186

# create the union table
union.keeps <- c("tumor.id", "type", "chr", "pos", "gene", "ref", "alt", "datasets", "combine")
calls.T_v_N.union <- unique(calls.T_v_N[, union.keeps])
calls.T_v_pooledN.union <- unique(calls.T_v_pooledN[, union.keeps])
calls.N_v_pooledN.union <- unique(calls.N_v_pooledN[, union.keeps])
calls.union <- unique(rbind(calls.T_v_N.union, calls.T_v_pooledN.union, calls.N_v_pooledN.union))

# label each variant based on what analysis it is identified by

# check if key is present in the individual datasets
calls.union <- calls.union %>% mutate(analyses = ifelse(combine %in% calls.T_v_N$combine, "T vs N", 
                           ifelse(combine %in% calls.T_v_pooledN$combine, "T vs pooledN", "N vs pooledN")))

# add more granular labelling
calls.union <- calls.union %>% 
  mutate(analyses = ifelse(combine %in% combine.T_v_N.T_v_pooledN, "T vs N, T vs pooledN", analyses))
calls.union <- calls.union %>% 
  mutate(analyses = ifelse(combine %in% combine.T_v_N.N_v_pooledN, "T vs N, N vs pooledN", analyses))
calls.union <- calls.union %>% 
  mutate(analyses = ifelse(combine %in% combine.T_v_pooledN.N_v_pooledN, "T vs pooledN, N vs pooledN", analyses))
calls.union <- calls.union %>% 
  mutate(analyses = ifelse(combine %in% combine.T_v_N.T_v_pooledN.N_v_pooledN, "T vs N, T vs pooledN, N vs pooledN", analyses))
calls.union <- unique(calls.union)

#to.merge <- c("tumor.depth", "tumor.var_reads", "tumor.allele_freq", "tumor.tissue_type", "tumor_content",
#              "effect", "impact", "biotype", "external_id", "combine")
to.merge <- c("tumor.depth", "tumor.var_reads", "tumor.allele_freq", "effect", "impact", "biotype", "external_id", "combine")

# add other fields depending on which tool
calls.T_v_N.to_merge <- unique(calls.T_v_N[, to.merge])
calls.T_v_pooledN.to_merge <- unique(calls.T_v_pooledN[, to.merge])
calls.N_v_pooledN.to_merge <- unique(calls.N_v_pooledN[, to.merge])

calls.T_v_N.analyses <- c("T vs N", "T vs N, T vs pooledN")
calls.T_v_pooledN.analyses <- c("T vs pooledN", "T vs pooledN, N vs pooledN")
calls.N_v_pooledN.analyses <- c("N vs pooledN")

calls.union.T_v_N.temp <- filter(calls.union, calls.union$analyses %in% calls.T_v_N.analyses)
calls.union.T_v_pooledN.temp <- filter(calls.union, calls.union$analyses %in% calls.T_v_pooledN.analyses)
calls.union.N_v_pooledN.temp <- filter(calls.union, calls.union$analyses %in% calls.N_v_pooledN.analyses)

# now add annotation to individual sets of analyses and combine again
calls.all.T_v_N <- merge(calls.union.T_v_N.temp, calls.T_v_N.to_merge, by = "combine")
calls.all.T_v_pooledN <- merge(calls.union.T_v_pooledN.temp, calls.T_v_pooledN.to_merge, by = "combine")
calls.all.N_v_pooledN <- merge(calls.union.N_v_pooledN.temp, calls.N_v_pooledN.to_merge, by = "combine")

calls.union <- rbind(calls.all.T_v_N, calls.all.T_v_pooledN, calls.all.N_v_pooledN)

# remove "datasets" columns since for some variants found by multiple analyses, the evidences are different
calls.union$datasets <- NULL

# add tumor content information
calls.union <- merge(calls.union, sample.data, by = "tumor.id", all.x = TRUE)

# sort the calls by case, chr, then position	
calls.union <- unique(arrange(calls.union, tumor.id, chr, pos))

#write.table(calls.union, "/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/calls.union.txt", sep = "\t", row.names = FALSE, quote = FALSE)
#write.table(calls.union, "/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/calls.union.germline_VAF_filtered.txt", sep = "\t", row.names = FALSE, quote = FALSE)
#write.table(calls.union, "/Volumes/shung/projects/gzl_targeted/blood_vs_pool_analysis/calls.union.germline_VAF_and_rsid_filtered.txt", sep = "\t", row.names = FALSE, quote = FALSE)

```

## ####################
## Post-union filters #
## ####################

## Apply ROC-based thresholds

```{r}
# Recall
#   if tumor content > 25%
#     filter for tumor variant reads >= 12
#   else (tumor content <= 25%) 
#     filter for tumor variant reads >= 8

# for all variants, filter for VAF >= 3%

MIN_VAR_READS_TUMOR.HIGH_TC = 12
MIN_VAR_READS_TUMOR.LOW_TC = 8
MIN_VAF = 3

calls.union <- filter(calls.union, 
                      (as.numeric(calls.union$tumor_content) > 25 & as.numeric(calls.union$tumor.var_reads >= 12)) |
                        (as.numeric(calls.union$tumor_content) <= 25 & as.numeric(calls.union$tumor.var_reads >= 8)))

calls.union <- filter(calls.union, as.numeric(calls.union$tumor.allele_freq) >= MIN_VAF)
```

## ###############
## Visualization #
## ###############
```{r}
library(ggplot2)
```

# Look at distributions of VAF for each type of analysis

```{r}
# temporarily remove "noisy" variants (VAF < 5%)
#calls.union.rsid_and_germline_VAF_filtered <- calls.union
#calls.union <- filter(calls.union, as.numeric(calls.union$tumor.allele_freq) > 5)
calls.union$analyses <- factor(calls.union$analyses, levels = c("T vs N", 
                                                                "T vs N, T vs pooledN", 
                                                                "T vs pooledN",
                                                                "T vs pooledN, N vs pooledN",
                                                                "N vs pooledN"))

p.violin <- ggplot(calls.union, 
             aes(analyses, as.numeric(tumor.allele_freq))) + 
  geom_violin(scale = "width", adjust = 0.5) +
  xlab("") +
  ylab("Tumor variant allele frequency (%)") +
  theme(legend.position="bottom") +
  theme(legend.title = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))

```

## Barplot showing overlap across samples

```{r}
# order samples by the number of mutations
num_mutations <- as.data.frame(table(calls.union$tumor.id))
num_mutations <- num_mutations[order(num_mutations$Freq, num_mutations$Var),]
num_mutations$Var1 <- factor(num_mutations$Var1)
calls.union$tumor.id <- factor(calls.union$tumor.id, levels = rev(num_mutations$Var1))

calls.union$analyses <- factor(calls.union$analyses, levels = c("T vs N", 
                                                                "T vs N, T vs pooledN", 
                                                                "T vs pooledN",
                                                                "T vs pooledN, N vs pooledN",
                                                                "N vs pooledN"))
p.bar <- ggplot(calls.union, aes(tumor.id)) + 
  geom_bar(aes(fill=analyses)) + 
  xlab("") +
  ylab("Number of mutations") +
  theme(legend.title = element_blank()) +
  theme(axis.text.x = element_text(angle = 45, hjust=1)) +
  scale_fill_manual(values = c("pink2","thistle3", "slategray3", "darkseagreen3", "palegreen2"))

```
