---	
title: "Intersect VarScan and Strelka"	
author: "Stacy Hung"	
date: "May 24, 2019"	
output: html_document	
---	
	
This script performs the following tasks:	
- reads in master file snv/indel results (summarized by canonical transcript and including batch and tumor content)
- plots distribution of tumor allele frequency per tumor sample, with information on tumor content
- plots #s of mutations per sample, stacked by variant class
- plots mutation frequency by panel gene (across total cohort) incl. non-mutated genes
 	
## Load libraries
 	
```{r}
library(dplyr)    # filter and rename
library(tidyr)    # separate	
```

## Create master list of SNVs/indels and add additional filters as needed

```{r}
calls.pilot <- read.table("/Volumes/shung/projects/gzl_targeted/pilot/calls.integration/calls-varscan_intersect_strelka-canonical.txt", sep = "\t", header = TRUE, fill = TRUE, stringsAsFactors = FALSE)

# remove "key" column
calls.pilot$key <- NULL

# create "plate" column (pilot, batch1, batch2, etc.)
calls.pilot$plate <- "pilot"

# merge with samples data to add columns for tumor content and normal method of extraction

# read in data that contains information on tumor content and tissue type of the normal
sample.data <- read.table("/Volumes/shung/projects/gzl_targeted/data/gzl_targeted-T_N-samples.txt", sep = "\t", header = TRUE, fill = TRUE)

# extract only required columns
sample.data <- sample.data[, c("tumor.target_ID", "tumor.tumor_id", "tumor.tissue_type", "tumor_content", "tumor.mean_target_coverage", "normal.extraction_method")]
colnames(sample.data) <- c("case_id", "tumor_id", "tumor.tissue_type", "tumor_content", "tumor.mean_target_coverage", "normal.extraction_method")

# merge T-N data with SNVs
calls.pilot <- merge(calls.pilot, sample.data, by = "tumor_id")

# remove the original "tumor_id" column (has long version of sample names) and rename "case id" to "tumor_id"
calls.pilot$tumor_id <- NULL
names(calls.pilot)[names(calls.pilot) == "case_id"] <- "tumor_id"

calls.batch1 <- read.table("/Volumes/shung/projects/gzl_targeted/targeted_batch1_20190812/calls.integration/calls-varscan_intersect_strelka.txt", sep = "\t", header = TRUE, fill = TRUE, stringsAsFactors = FALSE)

calls.batch1$plate <- "1"

# merge batches together
calls.pilot <- calls.pilot[colnames(calls.batch1)]
calls.all <- rbind(calls.pilot, calls.batch1)

# calls corresponding to pooled normals
calls.pooledN <- filter(calls.all, (calls.all$normal_id == "pooledN" | calls.all$normal_id == "BM1803125"))

# calls for matched tumor-normal pairs 
calls.matchedN <- filter(calls.all, calls.all$normal_id != "pooledN" & calls.all$normal_id != "BM1803125")
```


## Plot distribution of allele frequencies by tumor sample and include estimated tumor content

```{r}
library(gridExtra)
library(gtable)
library(ggplot2)

## AF distributions for tumors matched to a *POOLED* normal

# sort case_id by est_tumor_content
calls.pooledN <- calls.pooledN[order(calls.pooledN$tumor_content),]
calls.pooledN$tumor_id <- factor(calls.pooledN$tumor_id)

# for each sample, create AF density plot with vertical at peak value
AF.plots <- vector('list', length(unique(calls.pooledN$tumor_id)))
for (i in seq_along(unique(calls.pooledN$tumor_id))) {
  temp <- subset(calls.pooledN, calls.pooledN$tumor_id == unique(calls.pooledN$tumor_id)[i])
  
  num_mutations <- nrow(temp)
  
  # identify peak value for current sample
  max_y_density <- which.max(density(temp$allele_freq_tumor)$y)
  max_x_intercept <- density(temp$allele_freq_tumor)$x[max_y_density]
  max_y_intercept <- density(temp$allele_freq_tumor)$y[max_y_density]
  
  print(max_y_intercept)
  
  plot_label <- c(paste0("Peak AF: ", round(max_x_intercept, digits = 2), "% --", sep = ""))
  plot_label_2 <- c(paste0("\nEst. hetero. VAF --"))
  #plot_label_3 <- c(paste0("\n\nEst. germline VAF (T.C.) --"))
  
  AF.plots[[i]] <- ggplot(temp, aes(as.numeric(allele_freq_tumor))) + 
    geom_density(alpha = 0.2, fill="#FF6666") +
    geom_vline(xintercept = max_x_intercept, col="red", lty=2) +
    geom_vline(xintercept = as.numeric(temp$tumor_content / 2), col="blue", lty=2) +
    geom_vline(xintercept = 50, size = 20, alpha = 0.4, col="grey") +
    xlab("Allele frequency (%)") + ylab("Density") +
    theme(axis.title.x = element_blank(), axis.title.y = element_blank()) +
    ylim(0, max_y_intercept+0.25*max_y_intercept) +
    ggtitle(label = paste(temp$tumor_id, 
                          " (", temp$tumor.tissue_type, ", ",
                          round(as.numeric(factor(temp$tumor.mean_target_coverage[i])), 0), "X, ",
                          "n = ", num_mutations, ")", 
                          "; Est T.C. = ", unique(temp$tumor_content), "%", 
                          sep = "")) +
    annotate("text", x=Inf, y = Inf,
             label = plot_label, color = "red",
             vjust=1, hjust = 1) + 
    annotate("text", x=Inf, y = Inf,
             label = plot_label_2, color = "blue",
             vjust=1, hjust = 1)
}
#    geom_histogram(aes(y=..density..),      # Histogram with density instead of count on y-axis #binwidth=.5,
#                   binwidth=0.5,
#                   colour="black", fill="white") +

# landscape format (6 rows by 5 columns - printed at 20X11 pdf)
grid.arrange(grobs = AF.plots, ncol=5, bottom = "Allele frequency (%)", left = "Density")


## AF distributions for tumors with a *MATCHED* normal

# for some reason, calls done with the matched normal have been analyzed for all transcripts (not just canonical) -- need to remove transcript id and HGVS columns
calls.matchedN$transcript_id <- NULL
calls.matchedN$HGVS <- NULL
calls.matchedN <- unique(calls.matchedN)

# sort case_id by est_tumor_content
calls.matchedN <- calls.matchedN[order(calls.matchedN$tumor_content),]
calls.matchedN$tumor_id <- factor(calls.matchedN$tumor_id)

# for each sample, create AF density plot with vertical at peak value
AF.plots <- vector('list', length(unique(calls.matchedN$tumor_id)))
for (i in seq_along(unique(calls.matchedN$tumor_id))) {
  temp <- subset(calls.matchedN, calls.matchedN$tumor_id == unique(calls.matchedN$tumor_id)[i])
  
  num_mutations <- nrow(temp)
  
  # identify peak value for current sample
  max_y_density <- which.max(density(temp$allele_freq_tumor)$y)
  max_x_intercept <- density(temp$allele_freq_tumor)$x[max_y_density]
  max_y_intercept <- density(temp$allele_freq_tumor)$y[max_y_density]
  
  print(max_y_intercept)
  
  plot_label <- c(paste0("Peak AF: ", round(max_x_intercept, digits = 2), "% --", sep = ""))
  plot_label_2 <- c(paste0("\nEst. hetero. VAF --"))
  #plot_label_3 <- c(paste0("\n\nEst. germline VAF (T.C.) --"))
  
  AF.plots[[i]] <- ggplot(temp, aes(as.numeric(allele_freq_tumor))) + 
    geom_density(alpha = 0.2, fill="#FF6666") +
    geom_vline(xintercept = max_x_intercept, col="red", lty=2) +
    geom_vline(xintercept = as.numeric(temp$tumor_content / 2), col="blue", lty=2) +
    geom_vline(xintercept = 50, size = 20, alpha = 0.4, col="grey") +
    xlab("Allele frequency (%)") + ylab("Density") +
    annotate("text", x=Inf, y = Inf,
             label = plot_label, color = "red",
             vjust=1, hjust = 1) + 
    annotate("text", x=Inf, y = Inf,
             label = plot_label_2, color = "blue",
             vjust=1, hjust = 1) + 
    theme(axis.title.x = element_blank(), axis.title.y = element_blank()) +
    ylim(0, max_y_intercept+0.25*max_y_intercept) +
    ggtitle(label = paste(temp$tumor_id, 
                          " (", temp$tumor.tissue_type, ", ",
                          round(as.numeric(factor(temp$tumor.mean_target_coverage[i])), 0), "X, ",
                          "n = ", num_mutations, ")", 
                          "; Est T.C. = ", unique(temp$tumor_content), "%", 
                          sep = ""))
}
#    geom_histogram(aes(y=..density..),      # Histogram with density instead of count on y-axis #binwidth=.5,
#                   binwidth=0.5,
#                   colour="black", fill="white") +

# 16.5 X 9 pdf
grid.arrange(grobs = AF.plots, ncol=4, bottom = "Allele frequency (%)", left = "Density")

```

## Plot total numbers of mutations stacked by variant class

```{r}
library(ggplot2)
library(plyr) # revalue function

# make sure mutations are unique (based on a single transcript)
calls.all$transcript_id <- NULL
calls.all$HGVS <- NULL
calls.all <- unique(calls.all)

calls.all$class <- revalue(calls.all$effect, c("NON_SYNONYMOUS_CODING"="Missense", 
                           "NON_SYNONYMOUS_START"="Missense",
                           "START_GAINED"="Missense",
                           "START_LOST"="Start Lost",
                           "STOP_LOST"="Missense",
                           "STOP_GAINED"="Nonsense",
                           "STOP_GAINED+CODON_CHANGE_PLUS_CODON_DELETION"="Nonsense",
                           "CODON_CHANGE_PLUS_CODON_DELETION"="In-frame Indel",
                           "CODON_CHANGE_PLUS_CODON_INSERTION"="In-frame Indel",
                           "CODON_DELETION"="In-frame Indel",
                           "CODON_INSERTION"="In-frame Indel",
                           "FRAME_SHIFT"="Frameshift Indel",
                           "FRAME_SHIFT+START_LOST"="Frameshift Indel",
                           "SPLICE_SITE_ACCEPTOR"="Splice site",
                           "SPLICE_SITE_DONOR"="Splice site"
                           ))


calls.all$tumor_id <- factor(calls.all$tumor_id)
data.df <- as.data.frame(table(calls.all$tumor_id, calls.all$class))
colnames(data.df) <- c("case", "variant_class", "count")

# sort by total number of mutations
p <- ggplot(data = data.df, aes(x = reorder(case, -count), y = count, fill = variant_class)) + 
  geom_bar(stat = "identity", width = 0.6) + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) +
  scale_fill_discrete(name = "Variant Class") +
  xlab("") + ylab("Number of mutations") 
```


## Plot mutation frequencies for gene panel

```{r}
library(ggplot2)

# get gene panel
genes.panel <- read.table("/Volumes/shung/projects/gzl_exomes/data/target_panel_Agilent_genes.txt", header=FALSE)
genes.panel <- as.vector(genes.panel$V1)

# sort by decreasing number of mutations per gene
calls.all <- within(calls.all, gene <- factor(gene, levels=names(sort(table(gene), decreasing=FALSE))))

# plot
g <- ggplot(calls.all, aes(gene))
g + geom_bar(aes(), position = position_stack(reverse = TRUE)) +
  coord_flip() +
  theme(legend.position = "bottom")

# show recurrent genes (proportion of cases mutated per gene)

NUM_CASES = length(unique(calls.overlap$tumor_id))

# get absolute number of mutated cases
calls.unique_by_case <- unique(calls.overlap[c("tumor_id", "gene")])
gene.num_cases_mutated <- as.data.frame(table(calls.unique_by_case$gene))

# calculate proportion based on number of cases
gene.num_cases_mutated$proportion <- gene.num_cases_mutated$Freq / NUM_CASES * 100

colnames(gene.num_cases_mutated) <- c("gene", "freq", "proportion")

# write out to file for future referencing
write.table(gene.num_cases_mutated, "/Volumes/shung/projects/gzl_targeted/pilot/gene_mutation_frequencies-panel.txt", sep = "\t", row.names = FALSE, quote = FALSE)

# reorder genes by increasing frequency in PMBCL dataset
gene.num_cases_mutated$gene <- factor(gene.num_cases_mutated$gene, 
                                       levels = gene.num_cases_mutated$gene[order(gene.num_cases_mutated$proportion)])

p <- ggplot(data=gene.num_cases_mutated, aes(x = gene, y = proportion)) +
  geom_bar(stat="identity", position=position_dodge(), width = 0.8) +
  ylab("Cases mutated (%)") + xlab("") +
  theme_bw() +
  theme(legend.position="bottom") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size=8, vjust=0.5, face = "italic"),
        axis.text.y  = element_text(size=12), 
        axis.title.y = element_text(size = 14),
        plot.margin = unit(c(1,1,1,1), "cm"))
```

